### PR TITLE
NetworkController: Remove providerConfigChange event

### LINKED ISSUE #1205: NetworkController: Remove `NetworkController:providerConfigChange` controller event
This event is unnecessary for the following reasons:

1. There does not seem to be a reference to this event outside of the `core` codebase.
2. It is improperly named, as it is not fired immediately after the provider configuration object is changed, but rather after `network` in state is updated with the network ID.

### PR DESCRIPTION
**Pull request by mcmire**
## Description

Despite its name, the `providerConfigChange` event in NetworkController doesn't get emitted whenever the `providerConfig` property in state is changed. Rather, it gets updated as the very last step in `lookupNetwork`. This method is called:

- When the provider is initialized
- When the network is switched
- When the network is rolled back to the previous setting
- When `resetConnection` is called

`lookupNetwork` can also be called on its own.

If this were not enough, we already provide a way to listen for changes to a particular property in state: by passing a selector function to the `subscribe` method on the controller messenger.

So, this commit removes the `providerConfigChange` event from NetworkController. Instead of the following code:

```
messenger.subscribe(
  'NetworkController:providerConfigChange',
  (providerConfig) => {
    // do something with the providerConfig
  },
)
```

consumers should be able to say:

```
messenger.subscribe(
  'NetworkController:stateChange',
  (providerConfig) => {
    // do something with the providerConfig
  },
  (networkControllerState) => networkControllerState.providerConfig,
)
```

## Changes

- **BREAKING**: [`@metamask/network-controller`] Update `initializeProvider`, `lookupNetwork`, `setProviderType`, `setActiveNetwork`, `rollbackToPreviousProvider`, and `resetConnection` so that `NetworkController:providerConfigChange` is no longer emitted; remove this event entirely from the set of allowed events along with the `NetworkControllerProviderConfigChangeEvent` type
  - Consumers are encouraged to subscribe to `NetworkController:stateChange` with a selector function that returns `providerConfig` if they want to perform an action when `providerConfig` changes.
- **ADDED**: [`@metamask/network-controller`] Add `NetworkController:getState` to the set of allowed messenger actions, along with a `NetworkControllerGetStateAction` type
- **BREAKING:** [`@metamask/assets-controllers`] Update signature of `onNetworkStateChange` callback for TokenListController so that it must be given a NetworkController state change object; it will no longer also accept a provider config object
- **BREAKING:** [`@metamask/assets-controllers`] Update type of TokenListController messenger to require `NetworkController:stateChange` as an allowed event
- **BREAKING:** [`@metamask/gas-fee-controller`] Update type of GasFeeMessenger to require `NetworkController:getState` as an allowed event

## References

Fixes #1205.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/TokenListController.ts [hunk 1]
```diff
@@ -143,40 +141,30 @@ export class TokenListController extends BaseControllerV2<
...
-          await this.#onNetworkStateChangeCallback(
-            networkStateOrProviderConfig.providerConfig,
-          );
-        } else {
-          await this.#onNetworkStateChangeCallback(
-            networkStateOrProviderConfig,
-          );
-        }
+      onNetworkStateChange(async (networkControllerState) => {
+        await this.#onNetworkControllerStateChange(networkControllerState);
```
Comment (mcmire) [inline packages/assets-controllers/src/TokenListController.ts:145]: We _can_ simulate a network controller state object in tests, so we don't need to make this check.

File: packages/gas-fee-controller/src/GasFeeController.ts [hunk 2]
```diff
@@ -319,48 +317,30 @@ export class GasFeeController extends BaseControllerV2<
...
-        async (newProviderConfig) => {
-          this.ethQuery = this.messagingSystem.call(
-            'NetworkController:getEthQuery',
-          );
-
-          if (this.currentChainId !== newProviderConfig.chainId) {
-            this.currentChainId = newProviderConfig.chainId;
-            await this.resetPolling();
-          }
+        'NetworkController:stateChange',
```
Comment (mcmire) [inline packages/gas-fee-controller/src/GasFeeController.ts:341]: Note that we have a similar pattern of using `onNetworkStateChange` or subscribing to an event through the messenger in TokenListController, so I made them match each other.

File: packages/gas-fee-controller/src/GasFeeController.ts [hunk 3]
```diff
@@ -319,48 +317,30 @@ export class GasFeeController extends BaseControllerV2<
...
       getCurrentNetworkLegacyGasAPICompatibility;
-
     this.getCurrentAccountEIP1559Compatibility =
       getCurrentAccountEIP1559Compatibility;
+    this.#getProvider = getProvider;
     this.EIP1559APIEndpoint = EIP1559APIEndpoint;
     this.legacyAPIEndpoint = legacyAPIEndpoint;
     this.clientId = clientId;
+
+    this.ethQuery = new EthQuery(this.#getProvider());
```
Comment (mcmire) [inline packages/gas-fee-controller/src/GasFeeController.ts:329]: We don't need to call `getProvider` only if `onNetworkStateChange` and `getChainId` are given; this is a required option, so we can always get the provider this way. So I chose to use `getProvider` in the state change callback below to refresh the EthQuery instance. However, if we want to continue using the `getEthQuery` action we can — I just didn't know if we planned on getting rid of that or not.

**Review by Gudahtt — APPROVED** [review]
LGTM! I was concerned this would affect more controllers, but 2 is not bad. Nice that the gas fee controller and token list controller network change handling is now more concise and standardized.

I would suggest that the PR description be updated to include a note on the gas fee controller though; that change to the restricted messenger type is a breaking change as well.

**Comment by mcmire** [comment]
PR description updated and conflicts resolved.

**Review by Gudahtt — APPROVED** [review]
LGTM!

Though for:

>BREAKING: [@metamask/assets-controllers] Update type of messenger option for TokenListController to remove NetworkController:providerConfigChange from the set of allowed events

We should mention that the `NetworkController:stateChange` event is now required as well, since that's the breaking part. Removing an event is non-breaking.

**Review by Gudahtt — APPROVED** [review]

File: packages/gas-fee-controller/src/GasFeeController.ts [hunk 4]
```diff
@@ -319,48 +317,30 @@ export class GasFeeController extends BaseControllerV2<
...
       this.currentChainId = getChainId();
-      onNetworkStateChange(async () => {
-        const newProvider = getProvider();
-        const newChainId = getChainId();
-        this.ethQuery = new EthQuery(newProvider);
-        if (this.currentChainId !== newChainId) {
-          this.currentChainId = newChainId;
-          await this.resetPolling();
-        }
+      onNetworkStateChange(async (networkControllerState) => {
```
Comment (Gudahtt) [inline packages/gas-fee-controller/src/GasFeeController.ts:333]: Just noticed this breaking change that we missed. The state parameter wasn't used before, but now it is.

Comment (mcmire) [reply packages/gas-fee-controller/src/GasFeeController.ts:333]: Ah. That's a good point.

Comment (Gudahtt) [reply packages/gas-fee-controller/src/GasFeeController.ts:333]: Changelog updated here: https://github.com/MetaMask/core/pull/1405

**Pull request merged**